<launch>
    <!-- topic names -->
    <arg name="topic_imu"          default="/mpu9250/imu"/>
    <arg name="topic_mag"          default="/vn100t/magnetic_field"/>
    <!-- <arg name="topic_optitrack"    default="/optitrack/VN100T_CHK/poseStamped"/> -->
    <arg name="topic_optitrack"    default="/optitrack/BLUEFOX_MPU9250_CHK/poseStamped"/>

    <!-- topicnames of the 'nav_msgs' to be published -->
    <arg name="topic_nav_raw"      default="/state_estimator/raw"/>
    <arg name="topic_nav_filtered" default="/state_estimator/filtered"/>

    <!-- Verbose all estimation states at every second -->
    <arg name="verbose_all_estimation" default="true"/>
    <arg name="flag_visualizer_on"     default="true"/>

    <!-- IMU noise parameters -->
    <arg name="noise_accel"        default="0.004"/> <!-- Vectornav VN100t : 0.0008, MPU9250: 0.005-->
    <arg name="noise_gyro"         default="0.000005"/> <!-- Vectornav VN100t : 0.000005, MPU9250: 0.00005-->
    <arg name="noise_mag"          default="0.0001"/> <!-- no reference... not used in this algorithm-->

    <!-- IMU bias initial values. If you dont know, set zeros.-->
    <arg name="acc_bias"           default="[0.368753,0.217177,0.703889]"/> <!-- Vetornav VN100t: 0.011,0.007,0.201, MPU9250: 0,0,0-->
    <arg name="gyro_bias"          default="[-0.024338,0.0248388,0.0358478]"/> <!-- Vetornav VN100t: 0.00435,-0.0011758,-0.011671, MPU9250: 0,0,0-->
    <arg name="mag_bias"           default="[0.0,0.0,0.0]"/>

    <!-- Optitrack measurement covariance -->
    <arg name="noise_optitrack_position"    default="0.02"/>
    <arg name="noise_optitrack_orientation" default="0.01"/>

    <!-- Fixed rotation from the BODY frame (drone) to IMU frame -->
    <arg name="R_BI" default="[0.0,0.0,1.0,0.0,-1.0,0.0,1.0,0.0,0.0]"/> 
    <!-- Vectornav VN100t with Dongjae drone: [1.0,0.0,0.0,0.0,-1.0,0.0,0.0,0.0,-1.0]-->
    <!-- MPU9250 with CHK camera: [0.0,0.0,1.0,0.0,-1.0,0.0,1.0,0.0,0.0]-->

    <node pkg="state_estimator" name="state_estimator_node" type="state_estimator_node" output="screen">
        <param name="topic_imu"          type="string" value="$(arg topic_imu)"/>
        <param name="topic_mag"          type="string" value="$(arg topic_mag)"/>
        <param name="topic_optitrack"    type="string" value="$(arg topic_optitrack)"/>

        <param name="topic_nav_raw"      type="string" value="$(arg topic_nav_raw)"/>
        <param name="topic_nav_filtered" type="string" value="$(arg topic_nav_filtered)"/>
       
        <param name="verbose_all_estimation" type="bool" value="$(arg verbose_all_estimation)"/>
    
        <param name="noise_accel"        type="double" value="$(arg noise_accel)"/>
        <param name="noise_gyro"         type="double" value="$(arg noise_gyro)"/>
        <param name="noise_mag"          type="double" value="$(arg noise_mag)"/>
        
        <rosparam param="acc_bias"  subst_value="True"> $(arg acc_bias)  </rosparam>
        <rosparam param="gyro_bias" subst_value="True"> $(arg gyro_bias) </rosparam>
        <rosparam param="mag_bias"  subst_value="True"> $(arg mag_bias)  </rosparam>

        <param name="noise_optitrack_position"     type="double" value="$(arg noise_optitrack_position)"/>
        <param name="noise_optitrack_orientation"  type="double" value="$(arg noise_optitrack_orientation)"/>

       
        <rosparam param="R_BI"  subst_value="True"> $(arg R_BI)  </rosparam>
    </node>

    <group if="$(eval arg('flag_visualizer_on') ==true)">
        <node pkg="rviz" name="state_estimator_rviz" type="rviz" args="-d $(find state_estimator)/state_estimator.rviz"/>

        <node pkg="rqt_multiplot" name="state_estimator_multiplot" type="rqt_multiplot" output="screen">
            <param name="multiplot-config" type="string" value="$(find state_estimator)/state_estimator_plot.xml"/>
        </node>
    </group>

</launch>
